### ISSUE TITLE
Training stereo images, bad results

### ISSUE THREAD

**Issue by taiyar-046** [nianticlabs/monodepth2#192]
Dear authors, thanks for your great work. Here, I want to train the network using stereo images and inference using mono image. However, I found the training results was not good as your kitti model results. Following is the colored disparity result, could you please tell me what is the reason? Thank you again!
![0_disp](https://user-images.githubusercontent.com/65896848/82803409-0ba63d00-9eb3-11ea-9f51-ff473556c9b0.jpeg)

**Comment by mrharicot** [comment]
Hi,
It is impossible to help you without more details.
Please share a couple example training pairs.
How large is the training set?
Are the stereo images rectified?
Did you modify the intrinsics matrix `self.K` accordingly?
etc.

**Comment by taiyar-046** [comment]
Thank you for your reply. The settings of this project in my case is as following:
The original image size is width = 1280, and height = 720. 
K = (1191.603118, 0.000000, 640.530811, 0.000000, 1191.268535, 356.666798, 0.000000, 0.000000, 1.000000). 
Then the images are rectified. The rectified image size is width = 1191, height = 375. And the intrinsics after rectification is K = (1030.277, 0, 632.56, 0, 1030.277, 68.2. 0, 0, 1). 
So the final K which put in the program is [[0.865, 0, 0.53, 0],[0, 2.747, 0.1818, 0],[0, 0, 1, 0],[0, 0, 0, 1]].

**Comment by taiyar-046** [comment]
Following is an example of the rectification image pairs:
![left](https://user-images.githubusercontent.com/65896848/82851060-21f2de00-9f31-11ea-9e3a-73598c62f4df.jpg)
![right](https://user-images.githubusercontent.com/65896848/82851064-23bca180-9f31-11ea-8fb9-994f423b14d9.jpg)
And this is the reconstruction results:
![target](https://user-images.githubusercontent.com/65896848/82851117-5797c700-9f31-11ea-892c-5ddb7a819bd7.jpg)
![prediction](https://user-images.githubusercontent.com/65896848/82851120-59fa2100-9f31-11ea-93b0-92c38309d92c.jpg)

**Comment by mrharicot** [comment]
Hi,
It seems that the rectified stereo pair has negative disparities (check the top right of the image, it is going moving in the wrong direction), which means the rectified stereo cameras are not looking exactly in the same direction, there's a small angle between them.
The stereo training mode of monodepth2 assumes they are rectified with infinity at zero disparity (no negative disparities). You could try to train without the stereo flag and see if it helps.
There is also strong white balance difference between the two images which probably doesn't help training.
Is this a public dataset?

**Comment by taiyar-046** [comment]
Thanks for your advice. Cause I know the public dataset kitti_raw also has the negative disparity, you trained your model using that dataset, and has a good result. So how to explain about that? And you said monodepth2 assumes they are  rectified with infinity at zero disparity, what does that mean? Stereo camera has zero disparity?
This is not a public dataset, which made by me.

**Comment by mrharicot** [comment]
KITTI raw rectified has disparities in only one direction, which means that there is _no_ rotation between the rectified images, only a translation. This in turns results in mapping points at infinity to having zero disparity.
This isn't the case with your data and depending on their depth, points can either move left or right.
It is hard to say whether the white balance negatively affected the training, but it is very likely, as even with the right answer the loss will not be zero.

**Comment by taiyar-046** [comment]
Thanks for your reply. My recticified images also has disparity only in the u direction where it has the same value on the v direction, isn't it? I have checked several times to make sure they are in the same value in v direction.

**Comment by msceme** [comment]
> Thanks for your reply. My recticified images also has disparity only in the u direction while it has the same value on the v direction, isn't it? I have checked several times to make sure they are in the same value in v direction. Could you please tell me where you found the negative disparity in v direction in my case?

see the trees movement (top right)

**Comment by taiyar-046** [comment]
Thank you. I see that there is about 2 pixels difference in the v direction in top right of the images.

**Comment by mrharicot** [comment]
No I meant that the trees at the top right are moving in the opposite _horizontal_ direction as all the other pixels in the image, which means that you have both negative and positive disparities between the two images.

**Comment by taiyar-046** [comment]
Thanks a lot. There's a mistake when I did the rectification.

**Comment by taiyar-046** [comment]
Sorry to bother you again, cause I still got a not so good training result. The disparity result looks like this:
![1591766623185000000_disp](https://user-images.githubusercontent.com/65896848/84333194-9ed4b780-abc1-11ea-8f36-71a078eb29a6.jpeg)
and my training stereo images:
![1591778034208000000](https://user-images.githubusercontent.com/65896848/84333356-1c002c80-abc2-11ea-8fa4-9ea941f622be.jpg)
![right_10](https://user-images.githubusercontent.com/65896848/84333360-1f93b380-abc2-11ea-8bdd-789ad6099c87.jpg)
I trained the model for 30 epochs and there's 20000 images for thaining set and 2000 images for validation. Could you give me some advice for this problem? Thank you very much!

**Comment by CaptainEven** [comment]
@taiyar-046 May i ask how to  calculate the rectified image size after stereo rectification?

**Comment by anthonygofin** [comment]
> Thanks a lot. There's a mistake when I did the rectification.

Hi taiyar-046, 

Can I know which program you used for calibration and rectification ? I'm trying to train on custom dataset. I printed a chessboard and used OpenCV to calibrate and rectify my images, alas my rectified stereo images look weird and are unusable, so I am stuck and cannot perform training :(.

Thanks in advance for your reply.
